testutils/ethereum_source_data.go

```diff
@@ -0,0 +1,109 @@
+package testutils
+
+import (
+	"github.com/Zettablock/zsource/dao/ethereum"
+	"github.com/testcontainers/testcontainers-go/modules/postgres"
+	"gorm.io/gorm"
+)
+
+// These tables will be automatically created in each schema.
+var tableModels = []interface{}{
+	&ethereum.Block{},
+	&ethereum.Log{},
+	&ethereum.Trace{},
+	&ethereum.Transaction{},
+}
+
+// This struct is used to hold the data provided by the user for testing the
+// Ethereum handlers and initializing the database according to the data (by
+// creating tables and inserting data).
+type EthereumData struct {
+	// Map from schema name to the data for that schema.
+	data map[string]*EthereumSchemaData
+}
+
+func NewEthereumData() *EthereumData {
+	return &EthereumData{
+		data: make(map[string]*EthereumSchemaData),
+	}
+}
+
+func (d *EthereumData) AddSchemaData(schemaName string, blocks []*ethereum.Block, logs []*ethereum.Log) {
```

Comment:
what's the semantics? does all logs belong to a single block? or log to block is 1:1 mapping?

Reply:
there is no enforcement of semantics by the test runner other than the primary keys defined in the model. Caller can provide any data they think appropriate for the testing. So, the logs do not have to belong to a single block and there isn't any mapping between log and block required by the test runner.